Search flavors

gojamesgo85 opened the pull request:
Added flavors into the search view.

Because of the complexity of flavors being an optional array and expressions being so limited in their functionality, I placed the logic directly into complicated search. 

Trying to add the logic into makeCombinedPredicate was turning the code into an insanely verbose and lengthy function. 
I'm all ears if you have a better solution. But I'm happy with the result because we get to see that sweet loading wheel you made! 

I added all comments everywhere to explain my logic. 

I created a universal protocol for all ingredient enums to retrieve tags and reduce redundant conformances. 

Let me know what you think!

MatthewHunt84 commented:
Found a few bugs which need some fixin'

First up, if you add any category (umbrella, specialty, base) and then tap to remove it from the selections, the next time you search it the "included" stays on the cell which it shouldn't

Second, the generateComplicatedPredicates() function only gets called when the predicate is complicated. This means that the flavor logic only runs in cases where there are multiple categories being searched. If you just search "Gin" and make "angelica" unwanted, this doesn't qualify as a complicated search so none of the flavor logic gets executed and so a bunch of cocktails containing chartreuse will show up in the cocktail results.

gojamesgo85 commented:
Fixed thumbs cell issue. 
The main issue is that the cell were working form a different list than selectedIngredients. 
![Screenshot 2024-09-06 at 8 02 12 PM](https://github.com/user-attachments/assets/912a816d-e480-4993-9d4e-638a96492aa3)
We are caching all the category ingredients in updateCategoryIngredients, which are sets to remove duplicates. 

So the logic, which is the same in PreferencesThumbsCell to remove preferred selections, got copied and moved to "handleRemovalOf" func to remove the ingredients connected to categories from includedIngredientsSet if preferred == true and excludedIngredientsSet if preferred = false.
![Screenshot 2024-09-06 at 8 05 52 PM](https://github.com/user-attachments/assets/0809c18f-d985-4c8c-870b-df3f5e2a251c)

Finally, I aded conformance to evaluateSearchType for unwantedFlavors here:
![Screenshot 2024-09-06 at 8 06 47 PM](https://github.com/user-attachments/assets/524eb8f4-97f6-427a-8dfc-cc78045511c6)

gojamesgo85 commented:
Also, I changed the naming for the static var to find corresponding ingredients for their respective flavors so that it would read more clearly for its function.
![Screenshot 2024-09-06 at 8 13 01 PM](https://github.com/user-attachments/assets/38c3414e-eeb3-415c-895a-4a4a6a4492d2)

And my check for flavors is utilizing simpler syntax.
![Screenshot 2024-09-06 at 8 13 59 PM](https://github.com/user-attachments/assets/e63c988d-6839-4560-88e5-9bb578ce3d2a)
